File: backend/app/schemas/quality_analysis.py
from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, Field

DepthLevel = Literal["basic", "intermediate", "advanced"]
ConfidenceLevel = Literal["low", "medium", "high"]


class CoverageData(BaseModel):
    expected_keywords: list[str] = Field(default_factory=list)
    covered_keywords: list[str] = Field(default_factory=list)
    missing_keywords: list[str] = Field(default_factory=list)
    coverage_rate: float = Field(ge=0, le=1)
    explanation: str


class DepthAnalysis(BaseModel):
    expected_depth: DepthLevel
    actual_depth: DepthLevel
    score: float = Field(ge=0, le=1)
    explanation: str
    suggestions: list[str] = Field(default_factory=list)


class ConfidenceScore(BaseModel):
    level: ConfidenceLevel
    score: float = Field(ge=0, le=1)
    explanation: str
    factors: list[str] = Field(default_factory=list)


class EvidenceSource(BaseModel):
    source_type: str
    file_id: int | None = None
    knowledge_document_id: int | None = None
    chunk_id: str | None = None
    source_hash: str | None = None
    source_version: str | None = None
    retrieval_similarity: float | None = None
    reference_text: str


class KeypointMatch(BaseModel):
    keypoint: str
    evidence_chunk_id: str
    generated_section: str
    similarity: float = Field(ge=0, le=1)


class QualityAnalysis(BaseModel):
    analysis_version: str = "qa-v2"
    evidence_available: bool = False
    evidence_sources: list[EvidenceSource] = Field(default_factory=list)
    evidence_chunk_ids: list[str] = Field(default_factory=list)
    source_keypoints: list[str] = Field(default_factory=list)
    matched_keypoints: list[KeypointMatch] = Field(default_factory=list)
    missing_keypoints: list[str] = Field(default_factory=list)
    source_coverage: float | None = Field(default=None, ge=0, le=1)
    source_match_rate: float | None = Field(default=None, ge=0, le=1)
    diagnostic_confidence: float | None = Field(default=None, ge=0, le=1)
    constraint_fulfillment: float | None = Field(default=None, ge=0, le=1)
    warnings: list[str] = Field(default_factory=list)
    unavailable_reason: str | None = None
    algorithm: dict[str, Any] = Field(default_factory=dict)
    # qa-v1 compatibility. Historical snapshots remain readable without inventing qa-v2 values.
    coverage: CoverageData | None = None
    depth: DepthAnalysis | None = None
    confidence: ConfidenceScore | None = None
    suggestions: list[str] = Field(default_factory=list)

File: frontend/src/types/qualityAnalysis.ts
export type QualityDepthLevel = 'basic' | 'intermediate' | 'advanced'
export type QualityConfidenceLevel = 'low' | 'medium' | 'high'

export interface EvidenceSource {
  source_type: string
  file_id?: number | null
  knowledge_document_id?: number | null
  chunk_id?: string | null
  source_hash?: string | null
  source_version?: string | null
  retrieval_similarity?: number | null
  reference_text: string
}

export interface KeypointMatch {
  keypoint: string
  evidence_chunk_id: string
  generated_section: string
  similarity: number
}

export interface QualityAnalysis {
  analysis_version?: string
  evidence_available?: boolean
  evidence_sources?: EvidenceSource[]
  evidence_chunk_ids?: string[]
  source_keypoints?: string[]
  matched_keypoints?: KeypointMatch[]
  missing_keypoints?: string[]
  source_coverage?: number | null
  source_match_rate?: number | null
  diagnostic_confidence?: number | null
  constraint_fulfillment?: number | null
  warnings?: string[]
  unavailable_reason?: string | null
  algorithm?: Record<string, unknown>
  coverage?: { expected_keywords: string[]; covered_keywords: string[]; missing_keywords: string[]; coverage_rate: number; explanation: string }
  depth?: { expected_depth: QualityDepthLevel; actual_depth: QualityDepthLevel; score: number; explanation: string; suggestions: string[] }
  confidence?: { level: QualityConfidenceLevel; score: number; explanation: string; factors: string[] }
  suggestions?: string[]
}
